### PR TITLE
6257207: JTable.getDefaultEditor throws NullPointerException

### PR DESCRIPTION
**Pull request by prsadhuk**
If getDefaultEditor() is called before the JTable model is setup, it results in NPE.

This is because when JTable sets its model, which ends up firing a table changed event. The testcase is listening for tableChanged events and querying the editor. But the editor isn't installed until after the model is set which results in NPE.
Fix is to ensure initializeLocalVars() which initializes default editor is setup before JTable sets its model.

No regression is observed in jtreg/jck testsuite with this change.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-6257207](https://bugs.openjdk.org/browse/JDK-6257207): JTable.getDefaultEditor throws NullPointerException

### Reviewers
 * [Tejesh R](https://openjdk.org/census#tr) (@TejeshR13 - Committer) ⚠️ Review applies to [e2e5c38d](https://git.openjdk.org/jdk/pull/10871/files/e2e5c38d95d95c3850adfce132f28b16c6a7b5d0)
 * [Damon Nguyen](https://openjdk.org/census#dnguyen) (@DamonGuy - Author) ⚠️ Review applies to [e2e5c38d](https://git.openjdk.org/jdk/pull/10871/files/e2e5c38d95d95c3850adfce132f28b16c6a7b5d0)
 * [Alexey Ivanov](https://openjdk.org/census#aivanov) (@aivanov-jdk - **Reviewer**) ⚠️ Review applies to [a2c499fb](https://git.openjdk.org/jdk/pull/10871/files/a2c499fb995383f2dd16628f293aa0940e7dd305)
 * [Sergey Bylokhov](https://openjdk.org/census#serb) (@mrserb - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/10871/head:pull/10871` \
`$ git checkout pull/10871`

Update a local copy of the PR: \
`$ git checkout pull/10871` \
`$ git pull https://git.openjdk.org/jdk pull/10871/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 10871`

View PR using the GUI difftool: \
`$ git pr show -t 10871`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/10871.diff">https://git.openjdk.org/jdk/pull/10871.diff</a>

</details>

### CODE REVIEW

**Comment by TejeshR13** [comment]
The fix looks good to me and is working fine.

**Review by TejeshR13 — APPROVED** [review]

**Review by DamonGuy — APPROVED** [review]
The fix makes sense. Applied the fix and tested on my end. Looks good to me.

File: test/jdk/javax/swing/JTable/JTableBugTest.java [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+import javax.swing.JTable;
+import javax.swing.table.DefaultTableModel;
+import javax.swing.table.TableModel;
+import javax.swing.table.TableCellEditor;
+import javax.swing.table.TableCellRenderer;
+import javax.swing.event.TableModelEvent;
+import javax.swing.event.TableModelListener;
+import javax.swing.SwingUtilities;
+
+public class JTableBugTest {
```
Comment (aivanov-jdk) [inline test/jdk/javax/swing/JTable/JTableBugTest.java:41]: Can the test name be more specific?

File: test/jdk/javax/swing/JTable/JTableBugTest.java [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+        }
+
+        public void tableChanged(TableModelEvent event) {
+            super.tableChanged(event);
+
+            int column = event.getColumn();
+            if(column == TableModelEvent.ALL_COLUMNS)
+                resizeAll();
+            else
+                resize(column);
```
Comment (aivanov-jdk) [inline test/jdk/javax/swing/JTable/JTableBugTest.java:83]: Could you please update the code to follow the Java Code Style Guidelines? I mean, the braces, the spaces after keywords and opening parenthesis and around binary operators…

**Review by aivanov-jdk — CHANGES_REQUESTED** [review]

File: test/jdk/javax/swing/JTable/JTableEditorNPE.java [hunk 3]
```diff
@@ -77,39 +77,42 @@ public void tableChanged(TableModelEvent event) {
             super.tableChanged(event);
 
             int column = event.getColumn();
-            if(column == TableModelEvent.ALL_COLUMNS)
+            if (column == TableModelEvent.ALL_COLUMNS) {
                 resizeAll();
-            else
+            }
+            else {
```
Comment (aivanov-jdk) [inline test/jdk/javax/swing/JTable/JTableEditorNPE.java:83]:
```suggestion
            } else {
```

File: test/jdk/javax/swing/JTable/JTableEditorNPE.java [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+                myModel.addColumn("Col3");
+                myModel.addRow(new Object[]{"item1.1", "item1.2", "item1.3"});
+                myModel.addRow(new Object[]{"item2.1", "item2.2", "item2.3"});
+                MyTable tbl = new MyTable(myModel);
+                frame.getContentPane().add(tbl);
+                frame.setLocationRelativeTo(null);
+                frame.setVisible(true);
+            });
+        } finally {
+            Thread.sleep(1000);
```
Comment (aivanov-jdk) [inline test/jdk/javax/swing/JTable/JTableEditorNPE.java:62]: Is the `sleep` required?

If NPE is thrown, no UI appears at all; if NPE is not thrown, the UI may be shown but not required.

Would it be enough to add the table to the frame and call `validate()`?

File: test/jdk/javax/swing/JTable/JTableEditorNPE.java [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+                    if (comp != null) {
+                        width = Math.max(width, comp.getPreferredSize().width);
+                    }
+                }
+
+                renderer = this.getCellRenderer(row, column);
+                if (renderer != null) {
+                    comp = renderer.getTableCellRendererComponent(
+                            this, // table
+                            getModel().getValueAt(row,column), // value
```
Comment (aivanov-jdk) [inline test/jdk/javax/swing/JTable/JTableEditorNPE.java:118]:
```suggestion
                            getModel().getValueAt(row, column), // value
```

File: test/jdk/javax/swing/JTable/JTableEditorNPE.java (hunk 4 above)
Comment (prsadhuk) [reply test/jdk/javax/swing/JTable/JTableEditorNPE.java]: I added sleep to see the frame before it dispose but I agree it's not required for checking the fix..Modified as per comment...

**Review by aivanov-jdk — APPROVED** [review]

**Review by mrserb — APPROVED** [review]

**Pull request closed without merging**